### PR TITLE
Fixed Android Issue 834 - ConcurrentModificationException when try to…

### PR DESCRIPTION
**Pull request by hideki**
… update object

- Ticket: https://github.com/couchbase/couchbase-lite-android/issues/834

- Use `CopyOnWriteArraySet` or `CopyOnWriteArrayList` for Set/List of changeListener. CopyOnWriteArraySet/CopyOnWriteArrayList creates copy of array for iteration, so it does not cause `ConcurrentModificationException` and it is thread-safe.
- Except `changeListeners`, not applied CopyOnWriteArraySet/CopyOnWriteArrayList
- In addition, applied `final` for variables that should not be changed.

### CODE REVIEW

File: src/main/java/com/couchbase/lite/Database.java [hunk 1]
```diff
@@ -87,7 +88,7 @@
     public static int DEFAULT_MAX_REVS = 20;
 
     private Store store = null;
-    private String path;
+    final private String path;
```
Comment (pasin) [inline src/main/java/com/couchbase/lite/Database.java]: I think in the other classes, we use `private final`. It's better to have one style.

File: src/main/java/com/couchbase/lite/Document.java [hunk 2]
```diff
@@ -33,7 +34,7 @@
     /**
      * Change Listeners
      */
-    private List<ChangeListener> changeListeners = Collections.synchronizedList(new ArrayList<ChangeListener>());
+    final private List<ChangeListener> changeListeners = new CopyOnWriteArrayList<ChangeListener>();
```
Comment (pasin) [inline src/main/java/com/couchbase/lite/Document.java]: Same here. Using `private final` is more common, I guessed.

File: src/main/java/com/couchbase/lite/replicator/Replication.java [hunk 3]
```diff
@@ -77,7 +78,7 @@
     protected ScheduledExecutorService workExecutor;
     protected ReplicationInternal replicationInternal;
     protected Lifecycle lifecycle;
-    protected List<ChangeListener> changeListeners;
+    final private List<ChangeListener> changeListeners = new CopyOnWriteArrayList<ChangeListener>();
```
Comment (pasin) [inline src/main/java/com/couchbase/lite/replicator/Replication.java]: Same as above.

File: src/main/java/com/couchbase/lite/replicator/ReplicationInternal.java [hunk 4]
```diff
@@ -114,7 +115,7 @@
     protected ScheduledExecutorService workExecutor;
 
     protected StateMachine<ReplicationState, ReplicationTrigger> stateMachine;
-    protected List<ChangeListener> changeListeners;
+    final private List<ChangeListener> changeListeners = new CopyOnWriteArrayList<ChangeListener>();
```
Comment (pasin) [inline src/main/java/com/couchbase/lite/replicator/ReplicationInternal.java]: Same as above.

**Comment by hideki** [comment]
@pasin Please review this PR again.

**Pull request merged**